partiql-eval/src/plan.rs

```diff
@@ -29,17 +30,49 @@ use crate::eval::expr::{
 use crate::eval::EvalPlan;
 use partiql_value::Value::Null;
 
+#[macro_export]
+macro_rules! correct_num_args_or_err {
+    ($self:expr, $args:expr, $exact_num:literal, $name:expr) => {
+        if $args.len() != $exact_num {
+            $self.errors.push(EvaluationError::IllegalState(format!(
+                "Wrong number of arguments for {}",
+                $name.to_string()
+            )));
+            return Box::new(ErrorNode::new());
+        }
+    };
+    ($self:expr, $args:expr, $min_num:literal, $max_num:literal, $name:expr) => {
+        if !($min_num..=$max_num).contains(&$args.len()) {
+            $self
+                .errors
+                .push(EvaluationError::IllegalState($name.to_string()));
+            return Box::new(ErrorNode::new());
+        }
+    };
+}
+
 #[derive(Default)]
-pub struct EvaluatorPlanner;
+pub struct EvaluatorPlanner {
+    errors: Vec<EvaluationError>,
+}
 
 impl EvaluatorPlanner {
+    pub fn new() -> EvaluatorPlanner {
+        EvaluatorPlanner { errors: vec![] }
+    }
+
     #[inline]
-    pub fn compile(&self, plan: &LogicalPlan<BindingsOp>) -> EvalPlan {
-        self.plan_eval(plan)
+    pub fn compile(&mut self, plan: &LogicalPlan<BindingsOp>) -> Result<EvalPlan, EvalErr> {
+        if !self.errors.is_empty() {
+            return Err(EvalErr {
+                errors: self.errors.clone(),
+            });
+        }
+        Ok(self.plan_eval(plan))
```

Comment:
Planning is happening after the empty error check, but errors won't be populated until after plan.

```suggestion
        let plan = self.plan_eval(plan);
        if !self.errors.is_empty() {
            return Err(EvalErr {
                errors: std::mem::take(&mut self.errors),
            });
        }
        Ok(plan)
```